### PR TITLE
Display detailed error message if activation fails

### PR DESCRIPTION
**Pull request by malandrina**
* Currently when repo activation fails a generic error message
  is displayed to the user. This commit introduces support for
  displaying detailed, actionable error messages when repo activation fails.
* Update `RepoActivator` to add and expose errors when repo activation
  fails due to a GitHub permissions issue.
* Update `ActivationsController#create` to return error messages when
  error messages are present.
* Introduce `ErrorMessageTranslation` to handle translating
  non-user-friendly Octokit error messages into messages that can be
  displayed to the end user.
* Update client-side error handling to display error messages included
  in response body
* https://trello.com/c/KYuTdbXS/405-when-a-user-tries-to-enable-a-repo-and-it-fails-they-get-no-feedback-as-to-why-enabling-the-repo-failed

### CODE REVIEW

File: lib/error_message_translation.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# Parses OctoKit error messages
+# Returns human-readable error message
```
Comment (salbertson) [inline lib/error_message_translation.rb]: Hopefully this is clear based on the class and method names. Think we could remove these comments?

File: spec/lib/error_message_translation_spec.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
+require "spec_helper"
+
+describe ErrorMessageTranslation do
+  describe ".from_error_response" do
+    context "when error status is 403" do
+      it "returns error message" do
+        message = "PUT https://api.github.com/teams/3675/memberships/houndci: 403 - You must be an admin to add a team membership. // See: https://developer.github.com/v3"
```
Comment (salbertson) [inline spec/lib/error_message_translation_spec.rb]: What do you think about putting this in a fixture, similar to other responses?

**Comment by salbertson** [comment]
I like it! :+1: 

What error messages did you run across when testing? Just curious what errors we will see, for customer support.

File: app/controllers/activations_controller.rb [hunk 3]
```diff
@@ -7,12 +7,13 @@ class CannotActivatePaidRepo < StandardError; end
   before_action :check_repo_plan
 
   def create
-    if activator.activate
-      analytics.track_repo_activated(repo)
+    activator.activate
 
-      render json: repo, status: :created
+    if activator.errors.any?
+      render json: { errors: activator.errors }, status: 502
```
Comment (srt32) [inline app/controllers/activations_controller.rb]: Unrelated to this PR, but would it be interesting to track in analytics, failed repo activations?

Comment (malandrina) [inline app/controllers/activations_controller.rb]: That sounds like useful information. I made a card in the "Chores" list: https://trello.com/c/RKKrlrbB/487-track-failed-repo-activation-events

File: app/services/repo_activator.rb [hunk 4]
```diff
@@ -27,6 +30,7 @@ def activate_repo
   def change_repository_state_quietly
     yield
   rescue Octokit::Error => error
+    @errors.push(ErrorMessageTranslation.from_error_response(error))
```
Comment (srt32) [inline app/services/repo_activator.rb]: Can this be `errors`?

File: lib/error_message_translation.rb (hunk 1 above)
Comment (malandrina) [inline lib/error_message_translation.rb]: Sure, I'm happy to get rid of them. Removed in https://github.com/thoughtbot/hound/commit/811f34bf6e87e0a86620cedfa51543eb7e79fc32. I noticed we have descriptive comments for a bunch of classes in `app/models` and wasn't sure if we were still doing that.

File: spec/lib/error_message_translation_spec.rb (hunk 2 above)
Comment (malandrina) [inline spec/lib/error_message_translation_spec.rb]: Good call. It makes the tests much easier to read. Done in https://github.com/thoughtbot/hound/commit/811f34bf6e87e0a86620cedfa51543eb7e79fc32.

File: app/services/repo_activator.rb (hunk 4 above)
Comment (malandrina) [inline app/services/repo_activator.rb]: Yep! Good :eyes: . Done: https://github.com/thoughtbot/hound/commit/85c439aee3e8ffa557fbd030d6ca4262a47375ac

**Comment by malandrina** [comment]
@salbertson unfortunately I haven't found any GitHub documentation listing the error messages that their API might return. `"You must be an admin to add a team membership"` is the message I get when I try to activate a thoughtbot repo right now, so I figured I'd start there.

It looks like the Octokit gem builds its errors by looking at the GitHub API response and concatenating the request URL, status code, error message(s), and documentation URL ([reference](https://github.com/octokit/octokit.rb/blob/master/lib/octokit/error.rb)). More research is needed to figure out which GitHub error messages should be displayed to end users (and which ones shouldn't be).

File: app/assets/javascripts/directives/repo.js.coffee.erb [hunk 5]
```diff
@@ -6,9 +6,15 @@ App.directive 'repo', ['Subscription', 'StripeCheckout', 'User', (Subscription,
       scope.processing = true
 
       scope.repo.$activate()
-        .catch(-> alert('Your repo failed to activate.'))
+        .catch((response) -> alert(errorMessage(response.data)))
         .finally(-> scope.processing = false)
 
+    errorMessage = (responseData) ->
+      if responseData.errors
```
Comment (thorncp) [inline app/assets/javascripts/directives/repo.js.coffee.erb]: Will this always be truthy? It seems if `ErrorMessageTranslation` doesn't recognize the error, it will return nil, which gets pushed into the errors collection, resulting in `[nil]`, and serialized to json as `[null]`.

File: spec/support/fixtures/octokit_400_error.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+PUT https://api.github.com/teams/3675/memberships/houndci: 400 - Problems parsing JSON. // See: https://developer.github.com/v3
```
Comment (srt32) [inline spec/support/fixtures/octokit_400_error.txt]: We get plaintext errors back from Octokit?

File: app/assets/javascripts/directives/repo.js.coffee.erb (hunk 5 above)
Comment (salbertson) [inline app/assets/javascripts/directives/repo.js.coffee.erb]: What about this?

```
if responseData.errors.length > 0
...
```

File: spec/support/fixtures/octokit_400_error.txt (hunk 6 above)
Comment (malandrina) [inline spec/support/fixtures/octokit_400_error.txt]: We get an instance of `Octokit::Forbidden`, which has a `message` method. `message` returns a string. The string is a concatenation of the request URL, error response status, error message from the GitHub API, and the documentation URL.

File: app/assets/javascripts/directives/repo.js.coffee.erb (hunk 5 above)
Comment (malandrina) [inline app/assets/javascripts/directives/repo.js.coffee.erb]: @thorncp you're right -- in the case you describe the response is `"{ errors: [null] }"`. I'll work on changing that!

Comment (gylaz) [inline app/assets/javascripts/directives/repo.js.coffee.erb]:
```js
[null].length //=> 1
```
But we shouldn't be pushing that `nil` into the array.

File: app/controllers/activations_controller.rb [hunk 7]
```diff
@@ -25,7 +24,7 @@ def check_repo_plan
   end
 
   def activator
-    RepoActivator.new(repo: repo, github_token: github_token)
+    @activator ||= RepoActivator.new(repo: repo, github_token: github_token)
```
Comment (gylaz) [inline app/controllers/activations_controller.rb]: Why is this necessary?

File: lib/error_message_translation.rb [hunk 8]
```diff
@@ -0,0 +1,14 @@
+class ErrorMessageTranslation
```
Comment (gylaz) [inline lib/error_message_translation.rb]: Should this be `module`, to express that this isn't meant to be instantiated?

File: lib/error_message_translation.rb [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+  ERROR_CODE_AND_MESSAGE = /.*(\d{3})\s\-\s(.*)\.\s\/\/.*/
```
Comment (gylaz) [inline lib/error_message_translation.rb]: Good lord!

Comment (gylaz) [inline lib/error_message_translation.rb]: I'm not sure why, but I'm not getting match against:
```
GET https://api.github.com/repos/thoughtbot/houd: 404 - Not Found // See: https://developer.github.com/v3
```

Comment (gylaz) [inline lib/error_message_translation.rb]: I think it's because of us assuming there will be a dot `.`.

File: lib/error_message_translation.rb [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+  FORBIDDEN_ERROR_CODE = "403"
+
+  def self.from_error_response(error)
+    error_code_and_message = Hash.new
+
+    error.message.scan(ERROR_CODE_AND_MESSAGE) do |error_code, error_message|
+      error_code_and_message[error_code] = error_message
+    end
+
+    error_code_and_message.fetch(FORBIDDEN_ERROR_CODE, nil)
```
Comment (gylaz) [inline lib/error_message_translation.rb]: Wouldn't this method be simpler and be more readable as:
```
captures = error.message.match(ERROR_CODE_AND_MESSAGE).captures
if capture[0] == FORBIDDEN_ERROR_CODE
  captures[1]
end
```

File: spec/lib/error_message_translation_spec.rb [hunk 11]
```diff
@@ -0,0 +1,28 @@
+require "spec_helper"
```
Comment (gylaz) [inline spec/lib/error_message_translation_spec.rb]: I think this can be `fast_spec_helper`.

File: spec/support/fixtures/octokit_400_error.txt (hunk 6 above)
Comment (gylaz) [inline spec/support/fixtures/octokit_400_error.txt]: Any reason this is in a file, as opposed to just a string in the test?

File: app/services/repo_activator.rb [hunk 12]
```diff
@@ -27,6 +30,7 @@ def activate_repo
   def change_repository_state_quietly
     yield
   rescue Octokit::Error => error
+    errors.push(ErrorMessageTranslation.from_error_response(error))
```
Comment (gylaz) [inline app/services/repo_activator.rb]: How about throwing a `compact` on the end to strip out the `nil`s?

File: spec/support/fixtures/octokit_400_error.txt (hunk 6 above)
Comment (malandrina) [inline spec/support/fixtures/octokit_400_error.txt]: Scott and I thought it made the tests easier to read: https://github.com/thoughtbot/hound/pull/632#discussion_r24702535

File: app/controllers/activations_controller.rb (hunk 7 above)
Comment (malandrina) [inline app/controllers/activations_controller.rb]: `activator` is now being called twice: on line 10 and again on line 14.

File: lib/error_message_translation.rb (hunk 9 above)
Comment (malandrina) [inline lib/error_message_translation.rb]: Thanks for testing it out. I changed the regex in https://github.com/thoughtbot/hound/commit/7ff9921756e30e8454dabd5d67825313292288f0.

File: lib/error_message_translation.rb (hunk 8 above)
Comment (malandrina) [inline lib/error_message_translation.rb]: Sure, sounds good. Done: https://github.com/thoughtbot/hound/commit/7ff9921756e30e8454dabd5d67825313292288f0

File: lib/error_message_translation.rb (hunk 10 above)
Comment (malandrina) [inline lib/error_message_translation.rb]: Looks simpler to me. Done: https://github.com/thoughtbot/hound/commit/7ff9921756e30e8454dabd5d67825313292288f0

**Comment by malandrina** [comment]
@thorncp @salbertson @srt32 @gylaz ready for re-review.

**Comment by srt32** [comment]
More bubbles. I re-kicked the Travis job.

**Comment by malandrina** [comment]
@srt32 thanks again! We need a :bubble: emoji.

File: spec/support/fixtures/octokit_400_error.txt (hunk 6 above)
Comment (gylaz) [inline spec/support/fixtures/octokit_400_error.txt]: The only thing, it's not a response, but just a text representation of the error extracted from the response. I can see that confusing people.

**Comment by gylaz** [comment]
:+1:

File: spec/support/fixtures/octokit_400_error.txt (hunk 6 above)
Comment (malandrina) [inline spec/support/fixtures/octokit_400_error.txt]: @gylaz yeah, I agree that it might be confusing. What do you think about putting them in a subdirectory? e.g. `spec/support/fixtures/octokit/403_error.txt`

Comment (gylaz) [inline spec/support/fixtures/octokit_400_error.txt]: Personally, I feel like moving an arbitrary string to a file is a bit of a premature extraction. If it makes the test ugly because it's long, I'd extract it to a method first. Whatever, you feel is the best approach :smile:.

Comment (malandrina) [inline spec/support/fixtures/octokit_400_error.txt]: I think the fact that both you and @srt32 found it a little misleading/confusing is a good reason to not extract fixtures here. I'll make the change.

Comment (malandrina) [inline spec/support/fixtures/octokit_400_error.txt]: Done: https://github.com/thoughtbot/hound/commit/c81a0120614fe5147175b67d72d67647cf77f707
